docs/user-guide/setting-up-ci-cd.rst

```diff
@@ -0,0 +1,212 @@
+the performance of the NLU model without your changes to an NLU model with your
+changes. 
+
+You can do this by running NLU testing in cross-validation mode:
+
+.. code-block:: bash
+
+   rasa test nlu --cross-validation
+
+or by training a model on a training set and testing it on a test set. If you use the train-test
+set approach, it is best to :ref:`shuffle and split your data <train-test-split>` using ``rasa data split`` as part of this CI step, as
+opposed to using a static NLU test set, which can easily become outdated. 
+
+Since NLU comparison can be a fairly resource intensive test, you can set this
+test to run only when a certain tag (e.g. "NLU testing required") is present,
+or only when changes to NLU data or the NLU pipeline were made. 
+
+.. _testing-action-code:
+
+Test Action Code
+################
+
+The approach used to test your action code will depend on how it is
+implemented. Whichever method of testing your code you choose, you should
+include running those tests in your CI pipeline as well. For example, if you
+connect to external API's it is recommended to write unit tests to ensure 
+that those APIs respond as expected to common inputs.
+
+Continuous Deployment
+---------------------
+
+To get improvements out to your users frequently, you need to automate as
+much of the deployment process as possible. 
+
+CD steps usually run on push or merge to a certain branch, once CI checks have
+succeeded.
+
+.. contents::
+   :local:
+
+.. _uploading-a-model:
+
+Deploying your Rasa Model
+#########################
+
+You should already have a trained model from running end-to-end testing in your
+CI pipeline. You can set up your pipeline to upload the trained model to your
+Rasa server if the CI results are satisfactory. For example, in Rasa X this would look like:
+
+.. code-block:: bash
+
+   curl -k -F "model=@models/my_model.tar.gz" "https://example.rasa.com/api/projects/default/models?api_token={your_api_token}"
+
+If you are using Rasa X, you can also `tag the uploaded model <https://rasa.com/docs/rasa-x/api/rasa-x-http-api/#tag/Models/paths/~1projects~1{project_id}~1models~1{model}~1tags~1{tag}/put>`_
+as ``production`` (or whichever `deployment environment <https://rasa.com/docs/rasa-x/enterprise/deployment-environments/#>`_ 
+you want to deploy it to):
+
+.. code-block:: bash
+
+   curl -X PUT "https://example.rasa.com/api/projects/default/models/my_model/tags/production"
+
+
+However, if your update includes changes to both your model and your action
+code, and these changes depend on each other in any way, you should **not**
+automatically tag the model as ``production``. You will first need to build and
+deploy your updated action server, so that the new model won't e.g. call
+actions that don't exist in the pre-update action server.
+
+Deploying your Action Server
+############################
+
+If you're using a containerized deployment of your action server, you can
```

Comment:
How would you use it if you were not using a containerised deployment?

Reply:
I mean it's possible to just run it from a server without containerizing it but sure it's a pretty bad idea. I'll change it.